src/Compilers/VisualBasic/Test/Emit/Attributes/AttributeTests.vb

```diff
@@ -5111,5 +5113,456 @@ DiagID1: 'C' is for evaluation purposes only and is subject to change or removal
+<Module: System.Diagnostics.CodeAnalysis.Experimental("DiagID1")>
+
+Class Derived
+    Inherits C
+End Class
+]]>
+                                 </file>
+                             </compilation>
+
+            Dim derivedComp = CreateCompilation(derivedSrc, references:={originalC.ToMetadataReference(), attrRef})
+            derivedComp.AssertNoDiagnostics()
+
+            Dim comp = CreateCompilation("", references:={derivedComp.ToMetadataReference(), retargetedC.ToMetadataReference(), attrRef})
+            derivedComp.AssertNoDiagnostics()
+
+            Dim derivedType = comp.GetTypeByMetadataName("Derived")
+            Assert.IsType(Of RetargetingNamedTypeSymbol)(derivedType)
+            Assert.IsType(Of RetargetingModuleSymbol)(derivedType.ContainingModule)
+            Assert.Equal(ObsoleteAttributeKind.Experimental, derivedType.ContainingModule.ObsoleteKind)
+        End Sub
+
+        <Fact>
+        Public Sub MissingAssemblyAndModule()
+
+            Dim missingSrc = <compilation>
+                                 <file name="a.vb">
+                                     <![CDATA[
+Public Class Base
+End Class
+]]>
+                                 </file>
+                             </compilation>
+
+            Dim missingRef = CreateCompilation(missingSrc, assemblyName:="missing").EmitToImageReference()
+
+            Dim libSrc = <compilation>
+                             <file name="a.vb">
+                                 <![CDATA[
+Public Class C
+    Inherits Base
+
+    Public Shared Sub M()
+    End Sub
+End Class
+]]>
+                             </file>
+                         </compilation>
+
+            Dim libRef = CreateCompilation(libSrc, references:={missingRef}).EmitToImageReference()
+
+            Dim src = <compilation>
+                          <file name="a.vb">
+                              <![CDATA[
+Public Class Program
+    Public Shared Sub Main()
+        C.M()
+    End Sub
+End Class
+]]>
+                          </file>
+                      </compilation>
+
+            Dim comp = CreateCompilation(src, references:={libRef})
+
+            comp.AssertTheseDiagnostics(<errors><![CDATA[
+BC30652: Reference required to assembly 'missing, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null' containing the type 'Base'. Add one to your project.
+        C.M()
+        ~~~
+                                        ]]></errors>)
+
+            Dim missingType = comp.GlobalNamespace.GetTypeMember("C").BaseTypeNoUseSiteDiagnostics
+            Assert.True(TypeOf missingType.ContainingAssembly Is MissingAssemblySymbol)
+            Assert.Equal(ObsoleteAttributeKind.None, missingType.ContainingAssembly.ObsoleteKind)
+            Assert.True(TypeOf missingType.ContainingModule Is MissingModuleSymbol)
+            Assert.Equal(ObsoleteAttributeKind.None, missingType.ContainingModule.ObsoleteKind)
+        End Sub
+
+        <Fact>
+        Public Sub Cycle()
```

Comment:
>Cycle() [](http://example.com/codeflow?start=19&length=7)

Is there a corresponding C# test?

Reply:
Yes, all the C# tests that directly use the declaration of the attribute into the assembly that uses the attribute.
For example, `OnAssembly_UsedFromSource`